### PR TITLE
ML work

### PR DESCRIPTION
**Pull request by dayannat**

### CODE REVIEW

**Comment by jonathanm-12** [comment]
This is amazing work

We might want to try avoiding the usage of an external csv but since you were facing problems with data loss it's understandable. I'm sure we can experiment with many different approaches 

Maybe we can try exporting the data (within python) as a pandas table so that it can be parsed easier 

Kind of tricky to write, open, and read from a csv from multiple simultaneous processes 

I can try making a simple gui for this as well for ease of interaction

Goal is to implement this with brainflow, virtual env, and LSTM

**Review by jonathanm-12 — DISMISSED** [review]

**Comment by mitch-gilmore** [comment]
It looks like a good proof of concept and can definitely be in its current state used as an example implementation of using sockets, but given the way it is currently written (with lack of generality and extensibility) is not tenable to be embedded into the other processes. If this is the intention we can place it in documentation, otherwise additional effort will be required in order to improve its generality to be used in child processes. Please comment on the intention?

**Comment by jonathanm-12** [comment]
canceled my approval since this is still WIP, looking forward to see what you do with it

**Comment by dayannat** [comment]
This was meant to be a proof of concept for transferring data. That part is now working.

Still need to work on generalizing. Am currently exploring the PR made by Mitchell on 4/02 for getting data from the headset. Once that is working, will move onto the virtual environment.

**Review by jonathanm-12 — DISMISSED** [review]

File: EEG Classification (ML)/CNN.ipynb
Comment (mitch-gilmore) [inline EEG Classification (ML)/CNN.ipynb:1]: Can you please convert this to a python script? Jupyter notebooks tend to be a pain for VC integration since is save the outputs within the file. See line 149-152 for example. It would also make it easier to review.

File: EEG Classification (ML)/Train_Model.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                if verbose > 1:
+                    print('{} loss: {:.4f}, acc: {:.4f}'.format(phase,
+                                                                epoch_loss,
+                                                                epoch_acc))
+
+                # Deep copy the best model using early stopping
+                if phase == 'val' and epoch_acc > best_acc:
+                    best_acc = epoch_acc
+                    best_epoch = epoch
+                    best_model = copy.deepcopy(model)
```
Comment (mitch-gilmore) [inline EEG Classification (ML)/Train_Model.py:94]: This is not the standard way of taking a model snap shot and could cause problems if the model is in cuda memory. See [here](https://brsoff.github.io/tutorials/beginner/saving_loading_models.html#id3) for standard practices.

File: EEG Classification (ML)/Train_Model.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                loss = criterion(preds, labels.to(dev))
+
+                # Backpropagate & weight update
+                if phase == 'train':
+                    loss.backward()
+                    optimizer.step()
+                # store batch performance
+                with torch.no_grad():
+                    running_loss += float(loss.item())
+                    running_corrects += torch.sum(preds.data.max(1)[1] ==
```
Comment (mitch-gilmore) [inline EEG Classification (ML)/Train_Model.py:62]: Not sure if its a bug or not, but I have never seen a `data` member for Tensors. preds and labels are Tensors right? Also could not find anything in their documentation for Tensors about it.

**Comment by jonathanm-12** [comment]
Could you add comments in your code to indicate which parts of the work were written by you?

**Review by Bshortall — APPROVED** [review]

**Comment by mitch-gilmore** [comment]
@dayannat Please address my review comments. Additionally the the socket programming looks good. The ML code looks to be in a rough state (not in conformance with current PyTorch versions). In order to get your contributions in the main branch I would suggest making seperate PRs for the one for the socket programming and one for the ML. That way we can get the socket programming up streamed while we correct the ML code.

File: EEG Classification (ML)/Train_Model.py (hunk 2 above)
Comment (mitch-gilmore) [reply EEG Classification (ML)/Train_Model.py:62]: I found a link [here](https://stackoverflow.com/questions/51743214/is-data-still-useful-in-pytorch) explaining the `data` member. The overview is that using it currently shouldn't work if its a `Tensor`.

**Review by mitch-gilmore — CHANGES_REQUESTED** [review]
